src/item/FireworkRocketExplosion.php

```diff
@@ -0,0 +1,216 @@
+use pocketmine\data\SavedDataLoadingException;
+use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\utils\Binary;
+use pocketmine\utils\Utils;
+use function array_key_first;
+use function array_values;
+use function count;
+use function strlen;
+
+class FireworkRocketExplosion{
+
+	protected const TAG_TYPE = "FireworkType"; //TAG_Byte
+	protected const TAG_COLORS = "FireworkColor"; //TAG_ByteArray
+	protected const TAG_FADE_COLORS = "FireworkFade"; //TAG_ByteArray
+	protected const TAG_TWINKLE = "FireworkFlicker"; //TAG_Byte
+	protected const TAG_TRAIL = "FireworkTrail"; //TAG_Byte
+
+	public static function fromCompoundTag(CompoundTag $tag) : self{
+		$colors = self::decodeColors($tag->getByteArray(self::TAG_COLORS));
+		if(count($colors) === 0){
+			throw new SavedDataLoadingException("Colors list cannot be empty");
+		}
+
+		return new self(
+			FireworkRocketTypeIdMap::getInstance()->fromId($tag->getByte(self::TAG_TYPE)) ?? throw new SavedDataLoadingException("Invalid firework type"),
+			$colors,
+			self::decodeColors($tag->getByteArray(self::TAG_FADE_COLORS)),
+			$tag->getByte(self::TAG_TWINKLE, 0) !== 0,
+			$tag->getByte(self::TAG_TRAIL, 0) !== 0
+		);
+	}
+
+	/**
+	 * @return DyeColor[]
+	 */
+	protected static function decodeColors(string $colorsBytes) : array{
+		$colors = [];
+
+		$dyeColorIdMap = DyeColorIdMap::getInstance();
+		for($i = 0; $i < strlen($colorsBytes); $i++){
+			$color = $dyeColorIdMap->fromInvertedId(Binary::readByte($colorsBytes[$i]));
+			if($color !== null){
+				$colors[] = $color;
+			}else{
+				//TODO: should throw an exception?
+			}
+		}
+
+		return $colors;
+	}
+
+	/** @var \Closure(DyeColor) : void */
+	protected \Closure $colorsValidator;
+
+	/**
+	 * @param DyeColor[] $colors
+	 * @param DyeColor[] $fadeColors
+	 */
+	public function __construct(
+		protected FireworkRocketType $type,
+		protected array $colors,
+		protected array $fadeColors,
+		protected bool $twinkle,
+		protected bool $trail
+	){
+		$this->colorsValidator = function(DyeColor $_) : void{};
+
+		if(count($colors) === 0){
+			throw new \InvalidArgumentException("Colors list cannot be empty");
+		}
+		$this->setColors($colors);
+		$this->setFadeColors($fadeColors);
+	}
+
+	public function getType() : FireworkRocketType{
+		return $this->type;
+	}
+
+	/** @return $this */
+	public function setType(FireworkRocketType $type) : self{
+		$this->type = $type;
+		return $this;
+	}
+
+	/**
+	 * @return DyeColor[]
+	 */
+	public function getColors() : array{
+		return $this->colors;
+	}
+
+	/**
+	 * @param DyeColor[] $colors
+	 *
+	 * @return $this
+	 */
+	public function setColors(array $colors) : self{
+		if(count($colors) === 0){
+			throw new \InvalidArgumentException("Colors list cannot be empty");
+		}
+		Utils::validateArrayValueType($colors, $this->colorsValidator);
+		$this->colors = array_values($colors);
+
+		return $this;
+	}
+
+	/**
+	 * Returns the main color, this defines stuff like meta.
```

Comment:
What's the difference between "main colour" and the other colours? This documentation is really insufficient.